engine/servermaster/jobop/operator.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jobop
+
+import (
+	"go.uber.org/zap"
+	"golang.org/x/net/context"
+
+	"github.com/pingcap/log"
+	pkgOrm "github.com/pingcap/tiflow/engine/pkg/orm"
+	ormModel "github.com/pingcap/tiflow/engine/pkg/orm/model"
+	"github.com/uber-go/atomic"
+)
+
+type operateRouter interface {
+	SendCancelJobMessage(ctx context.Context, jobID string) error
+}
+
+// JobOperator abstracts a metastore based job operator, it encapsulates logic
+// to handle JobOp and a Tick API to ensure job moves towards to expected status.
+type JobOperator interface {
+	MarkJobCanceling(ctx context.Context, jobID string) error
+	MarkJobCanceled(ctx context.Context, jobID string) error
+	Tick(ctx context.Context) error
+}
+
+// JobOperatorImpl implements JobOperator
+type JobOperatorImpl struct {
+	frameMetaClient pkgOrm.Client
+	router          operateRouter
+	isRunning       atomic.Bool
+}
+
+// NewJobOperatorImpl creates a new JobOperatorImpl
+func NewJobOperatorImpl(cli pkgOrm.Client, router operateRouter) *JobOperatorImpl {
+	return &JobOperatorImpl{
+		frameMetaClient: cli,
+		router:          router,
+		isRunning:       *atomic.NewBool(false),
+	}
+}
+
+func (oper *JobOperatorImpl) setJobOperate(
+	ctx context.Context, jobID string, op ormModel.JobOpStatus,
+) error {
+	var ormFn func(ctx context.Context, JobID string) (pkgOrm.Result, error)
+	switch op {
+	case ormModel.JobOpStatusCanceling:
+		ormFn = oper.frameMetaClient.SetJobCanceling
+	case ormModel.JobOpStatusCanceled:
+		ormFn = oper.frameMetaClient.SetJobCanceled
+	default:
+		log.Panic("unexpected job operate", zap.Any("op", op))
+	}
+	if result, err := ormFn(ctx, jobID); err != nil {
+		return err
+	} else if result.RowsAffected() == 0 {
+		log.Info("job status is already set", zap.String("job-id", jobID), zap.Any("op", op))
+	}
+	return nil
+}
+
+// MarkJobCanceling implements JobOperator.MarkJobCanceling
+func (oper *JobOperatorImpl) MarkJobCanceling(ctx context.Context, jobID string) error {
+	return oper.setJobOperate(ctx, jobID, ormModel.JobOpStatusCanceling)
+}
+
+// MarkJobCanceled implements JobOperator.MarkJobCanceled
+func (oper *JobOperatorImpl) MarkJobCanceled(ctx context.Context, jobID string) error {
+	return oper.setJobOperate(ctx, jobID, ormModel.JobOpStatusCanceled)
+}
+
+// Tick implements JobOperator.Tick
+func (oper *JobOperatorImpl) Tick(ctx context.Context) error {
+	if !oper.isRunning.CAS(false, true) {
+		return nil
+	}
+	defer oper.isRunning.Swap(true)
```

Comment:
Why is this necessary?

Reply:
removed this logic. I was expect `Tick` can be called concurrently, but current it doesn't need